keps/sig-instrumentation/1668-log-tracking/README.md

```diff
@@ -0,0 +1,372 @@
+| spanid | We use SpanContext.SpanID as spanid<br>spanid spans a controller action.<br>spanid is unique for controller action |
+| initialtraceid | We implement new id(InitialTraceID) to SpanContext<br>We use [Baggage](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/baggage/api.md) to propagate initialtraceid<br/>We use UID of root object as initialtraceid<br>initialtraceid spans the entire object lifecycle. <br>initialtraceid is unique for related objects |
+
+All of three id's inception is from object creation and it dies with object deletion
+
+## Design Details
+
+### Prerequisite
+
+We need to consider three cases:
+- Case1: Requests from kubectl that creating an object
+- Case2: Requests from kubectl other than creating (e.g. updating, deleting) an object
+- Case3: Requests from controllers
+
+The design below is based on the above three cases
+
+The following picture show our design![design](./overview.png)
+
+
+
+we don't have any modifications in kubectl in this design.
+
+### API Server Tracing
+
+**1. refer to [KEP647](https://github.com/kubernetes/enhancements/issues/647)**
+
+KEP647 will help to
+
+```
+Goals
+The API Server generates and exports spans for incoming and outgoing requests.
+The API Server propagates context from incoming requests to outgoing requests.
+```
+
+**2. [Make up a new outgoing request](#design-of-id-progagation-in-client-go)**
+
+### Design of ID propagation (controller)
+
+**1. Propagate golang ctx from objects to API Calls**
+
+When controllers create/update/delete an object A based on another B, we propagate context from B to A. E.g.:
+```
+    ctx = httptrace.SpanContextFromAnnotations(context.Background(), objB.GetAnnotations())
+    err = r.KubeClient.CoreV1().Create(ctx, objA...)
+```
+We do propagation across objects without adding traces to that components.
+
+SpanContextFromAnnotations will do something like below pseudo code
+```
+func SpanContextFromAnnotations(){
+	// 1. decode annotation to SpanContext
+	// 2. generate Span from SpanContext
+	// 3. chain Span to ctx which is required by propagators.Inject
+}
+```
+**2. [Make up a new outgoing request](#design-of-id-progagation-in-client-go)**
+
+### Design of ID propagation in [client-go](https://github.com/kubernetes/client-go)
+
+[KEP647](https://github.com/kubernetes/enhancements/issues/647) will do this work too, so we can refer to it directly in the future.
+client-go helps to inject [TraceContext](https://pkg.go.dev/go.opentelemetry.io/otel/propagators#example-TraceContext) and [Baggage](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/baggage/api.md) to the outgoing http request header with API [RoundTrip](https://pkg.go.dev/go.opentelemetry.io/contrib/instrumentation/net/http#Transport.RoundTrip).
+
+apiserver and controllers use this API to make up a new outgoing request.
+
+### Design of Mutating webhook(Out of tree)
```

Comment:
I think this needs some introduction here.  Why do we need an admission controller at all?  Reasons I can think of include:

1. Ease of use.  Using client-go with a context.Context is easier than adding an annotation.  The webhook takes care of writing the annotation.
2. Object to object context propagation.  Without the mutating admission controller, we can only associate actions from a single object.  With the mutating admission controller, the logging metadata would be added for objects modified by controllers of the initial object (e.g. metadata added to a deployment annotation would appear in pod logs).